### PR TITLE
use generics instead of `any` in Option&Value

### PR DESCRIPTION
**Pull request by AlexErrant**
Apologies for the somewhat large PR; I couldn't figure out how to make it smaller.

### CODE REVIEW

File: src/create-select.tsx [hunk 1]
```diff
@@ -6,35 +6,32 @@ import {
   on,
 } from "solid-js";
 
-type Option = any;
+type Value<V> = V | V[];
 
-type SingleValue = any;
 
-type Value = SingleValue | SingleValue[];
```
Comment (AlexErrant) [inline src/create-select.tsx:13]: This here is the raison d'être; boo hiss `any`. Now this yields an error:

```tsx
<Select initialValue={1} options={["a", "b"]} optionToValue={(o) => { return new Date()}} />
// Type '(o: string) => Date' is not assignable to type '(option: string) => number'.
//  Type 'Date' is not assignable to type 'number'.
```

You can also do explicit types like

```tsx
<Select<string, string> initialValue={1} options={["a", "b"]} />
// Type 'number' is not assignable to type 'Value<string> | undefined'.
```

File: src/select.tsx [hunk 2]
```diff
@@ -31,25 +30,32 @@ interface CommonProps {
...
-type SelectReturn = ReturnType<typeof createSelect>;
+type SelectReturn<O,V> = ReturnType<typeof createSelect<O,V>>;
 
-type SelectProps = CreateSelectProps & Partial<CommonProps>;
+type SelectProps<O, V> = CreateSelectProps<O, V> & Partial<CommonProps<O,V>>;
 
-const SelectContext = createContext<SelectReturn>();
+function SelectContext<O,V>() {
+  return createContext<SelectReturn<O,V>>()
+};
```
Comment (AlexErrant) [inline src/select.tsx:42]: Note that this is a breaking change; it is now a function _returning_ a context. AFAIK only functions (and types) can have generics, so I had to wrap it in a fn.

File: src/create-select.tsx [hunk 3]
```diff
@@ -6,35 +6,32 @@ import {
...
 
-const createSelect = (props: CreateSelectProps) => {
+function createSelect<O, V>(props: CreateSelectProps<O, V>) {
   const config = mergeProps(
     {
       multiple: false,
       disabled: false,
-      optionToValue: (option: Option): SingleValue => option,
-      isOptionDisabled: (option: Option) => false,
+      optionToValue: (option: O): V => option as unknown as V,
```
Comment (AlexErrant) [inline src/create-select.tsx:28]: I'm relatively unhappy with this cast/default behavior... but better to describe the existing behavior than just to ignore the error. If I were a better dev I'd know how to make `optionToValue` non-optional if O doesn't extend V.

**Comment by AlexErrant** [comment]
Uh, please be aware I haven't tested _anything_; I just made the `raison d'être` changes and then fixed all the squigglies.

**Comment by martinpengellyphillips** [comment]
Looking at this again...I'm not sure the `raison d'être` is actually desired. It _is_ valid to return from `optionToValue` values of different types, because `format` can be used to handle that nicely. Unless I've misunderstood something - it has been a while since I looked at this :)

**Pull request closed without merging**